File: android/src/main/kotlin/com/example/easy_native/EasyNativePlugin.kt
package com.example.easy_native

import android.app.Activity
import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class EasyNativePlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    companion object {
        private var routerChannelRef: MethodChannel? = null
        private var eventChannel: MethodChannel? = null
        private var methodChannel: MethodChannel? = null

        fun emitToFlutter(type: String, data: Any? = null) {
            eventChannel?.invokeMethod(
                "emitToFlutter",
                mapOf("type" to type, "data" to data),
            )
        }

        fun invokeFlutter(method: String, data: Any? = null, result: Result? = null) {
            methodChannel?.invokeMethod(
                "invokeFlutter",
                mapOf("method" to method, "data" to data),
                result,
            )
        }

        fun completeRoute(
            requestId: String,
            resultValue: Any? = null,
            action: String = "nativeRouteComplete",
        ) {
            routerChannelRef?.invokeMethod(
                "completeRoute",
                mapOf(
                    "requestId" to requestId,
                    "result" to resultValue,
                    "success" to true,
                    "action" to action,
                ),
            )
        }
    }

    private lateinit var routerChannel: MethodChannel
    private var activity: Activity? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        routerChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "easy_native/router")
        routerChannel.setMethodCallHandler(this)
        routerChannelRef = routerChannel

        eventChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "easy_native/event_bus")
        eventChannel?.setMethodCallHandler(this)

        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "easy_native/methods")
        methodChannel?.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        EasyNativeLogger.log(EasyNativeLogLevel.DEBUG, "method call ${call.method}")
        when (call.method) {
            "isNativeRoute" -> {
                val routeName = call.argument<String>("routeName").orEmpty()
                result.success(EasyNativeRouteRegistry.isNativeRoute(routeName))
            }
            "hasActiveNativeFlow" -> result.success(EasyNativeFlowManager.hasActiveNativeFlow())
            "push" -> result.success(route(call, "push"))
            "replace" -> result.success(route(call, "replace"))
            "present" -> result.success(route(call, "present"))
            "pushAndRemoveUntil" -> result.success(route(call, "pushAndRemoveUntil"))
            "pop" -> result.success(EasyNativeFlowManager.pop(call.argument<Any?>("result")))
            "popUntil" -> {
                val routeName = call.argument<String>("routeName").orEmpty()
                result.success(EasyNativeFlowManager.popUntil(routeName))
            }
            "closeAll" -> result.success(EasyNativeFlowManager.closeAll(call.argument<Any?>("result")))
            "emitToNative" -> {
                val type = call.argument<String>("type").orEmpty()
                val data = call.argument<Any?>("data")
                EasyNative.handleNativeEvent(type, data)
                result.success(true)
            }
            "invokeNative" -> {
                val method = call.argument<String>("method").orEmpty().trim()
                val data = call.argument<Any?>("data")
                if (!EasyNative.handleNativeMethod(method, data, result)) {
                    result.notImplemented()
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        routerChannel.setMethodCallHandler(null)
        routerChannelRef = null
        eventChannel?.setMethodCallHandler(null)
        methodChannel?.setMethodCallHandler(null)
        eventChannel = null
        methodChannel = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        EasyNative.setup(binding.activity.applicationContext)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        EasyNative.setup(binding.activity.applicationContext)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    private fun route(call: MethodCall, action: String): Map<String, Any?> {
        val routeName = call.argument<String>("routeName").orEmpty()
        val arguments = call.argument<Any?>("arguments")
        val requestId = call.argument<String>("requestId")
        val context = currentContext()
            ?: return mapOf(
                "success" to false,
                "action" to action,
                "message" to "No Android context is available",
            )

        return when (action) {
            "push" -> EasyNativeFlowManager.push(context, routeName, arguments, requestId)
            "replace" -> {
                if (EasyNativeFlowManager.hasActiveNativeFlow()) {
                    EasyNativeFlowManager.replace(
                        context,
                        routeName,
                        arguments,
                        call.argument<Any?>("result"),
                        requestId,
                    )
                } else {
                    EasyNativeFlowManager.push(context, routeName, arguments, requestId)
                }
            }
            "present" -> EasyNativeFlowManager.present(context, routeName, arguments, requestId)
            "pushAndRemoveUntil" -> EasyNativeFlowManager.pushAndRemoveUntil(
                context,
                routeName,
                arguments,
                call.argument<String>("untilRoute"),
                requestId,
            )
            else -> mapOf(
                "success" to false,
                "action" to action,
                "message" to "Unknown route action",
            )
        }
    }

    private fun currentContext(): Context? {
        return activity
    }
}

File: ios/easy_native/Sources/easy_native/EasyNativePlugin.swift
import Flutter
import UIKit

public final class EasyNative {
    public static let shared = EasyNative()

    public var rootNavigatorProvider: (() -> UINavigationController?)?

    private init() {}

    public func setup(rootNavigatorProvider: @escaping () -> UINavigationController?) {
        self.rootNavigatorProvider = rootNavigatorProvider
    }

    public func setLogProvider(_ provider: EasyNativeLogger.Provider?) {
        EasyNativeLogger.provider = provider
    }

    public func registerNativeRoute(_ name: String, factory: @escaping (Any?) -> UIViewController?) {
        EasyNativeRouteRegistry.shared.registerNativeRoute(name, factory: factory)
    }

    public func registerNativeMethod(
        _ method: String,
        handler: @escaping (Any?, @escaping FlutterResult) -> Void
    ) {
        EasyNativePlugin.shared?.registerNativeMethod(method, handler: handler)
    }

    public func registerNativeEventHandler(_ handler: @escaping (String, Any?) -> Void) {
        EasyNativePlugin.shared?.registerNativeEventHandler(handler)
    }

    public func push(_ routeName: String, arguments: Any? = nil) -> [String: Any] {
        return EasyNativeFlowManager.shared.push(
            routeName: routeName,
            arguments: arguments,
            from: rootNavigatorProvider?(),
            requestId: nil
        )
    }

    public func replace(_ routeName: String, arguments: Any? = nil) -> [String: Any] {
        if EasyNativeFlowManager.shared.hasActiveNativeFlow {
            return EasyNativeFlowManager.shared.replace(
                routeName: routeName,
                arguments: arguments,
                result: nil,
                requestId: nil
            )
        }
        return EasyNativeFlowManager.shared.push(
            routeName: routeName,
            arguments: arguments,
            from: rootNavigatorProvider?(),
            requestId: nil
        )
    }

    public func present(_ routeName: String, arguments: Any? = nil) -> [String: Any] {
        return EasyNativeFlowManager.shared.present(
            routeName: routeName,
            arguments: arguments,
            from: rootNavigatorProvider?(),
            requestId: nil
        )
    }

    public func pop(result: Any? = nil) -> [String: Any] {
        return EasyNativeFlowManager.shared.pop(result: result)
    }

    public func popUntil(_ routeName: String) -> [String: Any] {
        return EasyNativeFlowManager.shared.popUntil(routeName: routeName)
    }

    public func closeAll(result: Any? = nil) -> [String: Any] {
        return EasyNativeFlowManager.shared.closeAll(result: result)
    }
}

public final class EasyNativePlugin: NSObject, FlutterPlugin {
    private var routerChannel: FlutterMethodChannel?
    private var eventChannel: FlutterMethodChannel?
    private var methodChannel: FlutterMethodChannel?
    private var nativeMethodHandlers: [String: (Any?, @escaping FlutterResult) -> Void] = [:]
    private var nativeEventHandlers: [(String, Any?) -> Void] = []

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = EasyNativePlugin()
        instance.routerChannel = FlutterMethodChannel(
            name: "easy_native/router",
            binaryMessenger: registrar.messenger()
        )
        instance.eventChannel = FlutterMethodChannel(
            name: "easy_native/event_bus",
            binaryMessenger: registrar.messenger()
        )
        instance.methodChannel = FlutterMethodChannel(
            name: "easy_native/methods",
            binaryMessenger: registrar.messenger()
        )

        registrar.addMethodCallDelegate(instance, channel: instance.routerChannel!)
        registrar.addMethodCallDelegate(instance, channel: instance.eventChannel!)
        registrar.addMethodCallDelegate(instance, channel: instance.methodChannel!)
        EasyNativePlugin.shared = instance
    }

    fileprivate static weak var shared: EasyNativePlugin?

    public static func emitToFlutter(type: String, data: Any? = nil) {
        shared?.eventChannel?.invokeMethod("emitToFlutter", arguments: [
            "type": type,
            "data": data ?? NSNull(),
        ])
    }

    public static func invokeFlutter(method: String, data: Any? = nil, result: FlutterResult? = nil) {
        shared?.methodChannel?.invokeMethod(
            "invokeFlutter",
            arguments: [
                "method": method,
                "data": data ?? NSNull(),
            ],
            result: result
        )
    }

    public static func completeRoute(
        requestId: String,
        result: Any? = nil,
        action: String = "nativeRouteComplete"
    ) {
        shared?.routerChannel?.invokeMethod("completeRoute", arguments: [
            "requestId": requestId,
            "result": result ?? NSNull(),
            "success": true,
            "action": action,
        ])
    }

    public func registerNativeMethod(
        _ method: String,
        handler: @escaping (Any?, @escaping FlutterResult) -> Void
    ) {
        let methodName = method.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !methodName.isEmpty else {
            EasyNativeLogger.log(.warning, "ignore empty native method registration")
            return
        }
        if nativeMethodHandlers[methodName] != nil {
            EasyNativeLogger.log(.warning, "override native method registration \(methodName)")
        }
        nativeMethodHandlers[methodName] = handler
    }

    public func registerNativeEventHandler(_ handler: @escaping (String, Any?) -> Void) {
        nativeEventHandlers.append(handler)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        EasyNativeLogger.log(.debug, "method call \(call.method)")
        switch call.method {
        case "isNativeRoute":
            let args = call.arguments as? [String: Any] ?? [:]
            result(EasyNativeRouteRegistry.shared.isNativeRoute(args["routeName"] as? String ?? ""))
        case "hasActiveNativeFlow":
            result(EasyNativeFlowManager.shared.hasActiveNativeFlow)
        case "push":
            result(route(arguments: call.arguments, action: .push))
        case "replace":
            result(route(arguments: call.arguments, action: .replace))
        case "present":
            result(route(arguments: call.arguments, action: .present))
        case "pop":
            let args = call.arguments as? [String: Any] ?? [:]
            result(EasyNativeFlowManager.shared.pop(result: args["result"]))
        case "popUntil":
            let args = call.arguments as? [String: Any] ?? [:]
            result(EasyNativeFlowManager.shared.popUntil(routeName: args["routeName"] as? String ?? ""))
        case "pushAndRemoveUntil":
            result(route(arguments: call.arguments, action: .pushAndRemoveUntil))
        case "closeAll":
            let args = call.arguments as? [String: Any] ?? [:]
            result(EasyNativeFlowManager.shared.closeAll(result: args["result"]))
        case "emitToNative":
            let args = call.arguments as? [String: Any] ?? [:]
            let type = args["type"] as? String ?? ""
            let data = args["data"]
            nativeEventHandlers.forEach { $0(type, data) }
            result(true)
        case "invokeNative":
            let args = call.arguments as? [String: Any] ?? [:]
            let method = (args["method"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if let handler = nativeMethodHandlers[method] {
                handler(args["data"], result)
            } else {
                result(FlutterMethodNotImplemented)
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private enum NativeRouteAction {
        case push
        case replace
        case present
        case pushAndRemoveUntil
    }

    private func route(arguments: Any?, action: NativeRouteAction) -> [String: Any] {
        let args = arguments as? [String: Any] ?? [:]
        let routeName = args["routeName"] as? String ?? ""
        switch action {
        case .push:
            return EasyNativeFlowManager.shared.push(
                routeName: routeName,
                arguments: args["arguments"],
                from: EasyNative.shared.rootNavigatorProvider?(),
                requestId: args["requestId"] as? String
            )
        case .replace:
            if EasyNativeFlowManager.shared.hasActiveNativeFlow {
                return EasyNativeFlowManager.shared.replace(
                    routeName: routeName,
                    arguments: args["arguments"],
                    result: args["result"],
                    requestId: args["requestId"] as? String
                )
            }
            return EasyNativeFlowManager.shared.push(
                routeName: routeName,
                arguments: args["arguments"],
                from: EasyNative.shared.rootNavigatorProvider?(),
                requestId: args["requestId"] as? String
            )
        case .present:
            return EasyNativeFlowManager.shared.present(
                routeName: routeName,
                arguments: args["arguments"],
                from: EasyNative.shared.rootNavigatorProvider?(),
                requestId: args["requestId"] as? String
            )
        case .pushAndRemoveUntil:
            if EasyNativeFlowManager.shared.hasActiveNativeFlow {
                return EasyNativeFlowManager.shared.pushAndRemoveUntil(
                    routeName: routeName,
                    arguments: args["arguments"],
                    untilRoute: args["untilRoute"] as? String,
                    requestId: args["requestId"] as? String
                )
            }
            return EasyNativeFlowManager.shared.push(
                routeName: routeName,
                arguments: args["arguments"],
                from: EasyNative.shared.rootNavigatorProvider?(),
                requestId: args["requestId"] as? String
            )
        }
    }
}
